### DISCUSSION TITLE
Magick++ map() api does not support FloydSteinberg dither option

### DISCUSSION THREAD

**Discussion by filippobordon** [ImageMagick/ImageMagick#6017] [Development]
### ImageMagick version

7.1.0.60

### Operating system, version and so on

multi arch

### Description

Hi,
I have started using magick++ api to replace convert util, and in my scenario I am using convert as following:

convert input.jpeg -resize 400x300^ -dither FloydSteinberg -define dither:diffusion-amount=90% -remap pattern:gray50  BMP3:"output.bmp"

image.map(remap, true);
It seems to me that there is no option to remap an image with FloydSteinberg dither, indeed map(..., bool dither) automatically selects RiemersmaDitherMethod when set to true.
how about letting the user select the dither method, I,e, changing the map api to map(..., const DitherMethod ditherMethod_) ?

**Comment by snibgo** [comment]
How about ...
```
void Magick::Image::quantizeDitherMethod(const DitherMethod ditherMethod_)
```
... in Image.cpp?

**Reply by filippobordon** [reply]
I have tried already that approach, but it does not achieve the expected result of that convert command.
The convert command above does a remap applying also a FloydSteinberg dither to the image.

I tried i.e.
`image.artifact("dither:diffusion-amount", "90%");
image.quantizeDitherMethod(MagickCore::DitherMethod::FloydSteinbergDitherMethod);
image.quantizeColors(256);
image.quantize();
image.map(remap, false);`
but does not return the same image as that convert command.
While if I modify map() to use FloydSteinbergDitherMethod in case dither is set to true and I execute the following code:

`image.artifact("dither:diffusion-amount", "90%");
image.map(remap, true); // Modified to use `FloydSteinberg`

 then I get the same output as the convert above.

**Reply by dlemstra** [reply]
I can confirm that you currently can only use `RiemersmaDitherMethod` when doing `image.map`. And I am not yet sure how we should make this possible.

**Comment by psiegl** [comment]
Stumbled accross the same issue. And until today, this is not appropriately fixed. However, it seems there is a workaround:

The methods of interest are as follows:

```
// file lib/Options.cpp
// the official quantizeDither, sets dither as True and subsequently Riemersma
void Magick::Options::quantizeDither(const bool ditherFlag_)
{
  _imageInfo->dither=(MagickBooleanType) ditherFlag_;
  _quantizeInfo->dither_method=ditherFlag_ ? RiemersmaDitherMethod :
    NoDitherMethod;
}

// the more open one does not set dither True, but at least we can set FloydSteinberg, thus combine with above.
void Magick::Options::quantizeDitherMethod(const DitherMethod ditherMethod_)
{
  _quantizeInfo->dither_method=ditherMethod_;
}

// file lib/Image.cpp
// solely of interest in case of duplicate image, skip for now.
void Magick::Image::modifyImage(void)
{
  if (!_imgRef->isShared())
    return;

  GetPPException;
  replaceImage(CloneImage(image(),0,0,MagickTrue,exceptionInfo));
  ThrowImageException;
}

void Magick::Image::map(const Image &mapImage_,const bool dither_)
{
  modifyImage();
  GetPPException;
  options()->quantizeDither(dither_);
  RemapImage(options()->quantizeInfo(),image(),mapImage_.constImage(),
    exceptionInfo);
  ThrowImageException;
}
```
So with above inline comments, one can come up with the following (one could make it nicer ...):

```
void map_FloydSteinbergDitherMethod(Magick::Image& img, const Magick::Image &mapImage_ /*,const bool dither_*/)
{
  //modifyImage();
  //GetPPException;
  img.quantizeDither(true /* dither_ */);
  img.quantizeDitherMethod(MagickCore::FloydSteinbergDitherMethod);  // Use Floyd-Steinberg dithering
  RemapImage(img.quantizeInfo(),img.image(),mapImage_.constImage(), MagickCore::AcquireExceptionInfo());

  //ThrowImageException;
}
```

**Reply by psiegl** [reply]
https://github.com/ImageMagick/ImageMagick/pull/7937
